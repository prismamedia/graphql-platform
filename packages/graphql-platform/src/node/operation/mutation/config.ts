import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
