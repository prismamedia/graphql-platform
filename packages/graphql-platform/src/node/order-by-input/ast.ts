export interface ISortValue<TKind extends string> {
  readonly kind: TKind;
  readonly direction: 'ASC' | 'DESC';
}

export interface ILeafSortValue extends ISortValue<'Leaf'> {
  readonly leaf: string;
}

export type TSortValue = ILeafSortValue;
