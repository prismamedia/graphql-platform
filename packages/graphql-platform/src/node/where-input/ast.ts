import { TLeafValue } from '../component/leaf';

interface IFilterValue<TKind extends string, TValue> {
  readonly kind: TKind;
  readonly value: TValue;
}

export type TBooleanFilterValue = IFilterValue<'Boolean', boolean>;

interface ILogicalFilterValue<TOperator extends string, TValue>
  extends IFilterValue<'Logical', TValue> {
  readonly operator: TOperator;
}

export type TLogicalFilterValue =
  | ILogicalFilterValue<'and', ReadonlyArray<TFilterValue>>
  | ILogicalFilterValue<'or', ReadonlyArray<TFilterValue>>
  | ILogicalFilterValue<'not', TFilterValue>;

interface ILeafFilterValue<TOperator extends string, TValue>
  extends IFilterValue<'Leaf', TValue> {
  readonly leaf: string;
  readonly operator: TOperator;
}

export type TLeafFilterValue<TValue extends TLeafValue = any> =
  | ILeafFilterValue<'eq', TValue>
  | ILeafFilterValue<'not', TValue>
  | ILeafFilterValue<'gt', TValue>
  | ILeafFilterValue<'gte', TValue>
  | ILeafFilterValue<'lt', TValue>
  | ILeafFilterValue<'lte', TValue>
  | ILeafFilterValue<'in', ReadonlyArray<TValue>>
  | ILeafFilterValue<'not_in', ReadonlyArray<TValue>>;

export type TLeafFilterOperator<
  T extends TLeafFilterValue['operator'] = TLeafFilterValue['operator']
> = Extract<TLeafFilterValue['operator'], T>;

export const isLeafFilterOperatorAmong = (
  maybeOperator: unknown,
  operators: ReadonlyArray<TLeafFilterOperator>,
): maybeOperator is TLeafFilterOperator =>
  operators.includes(maybeOperator as any);

interface IEdgeFilterValue<TOperator extends string, TValue>
  extends IFilterValue<'Edge', TValue> {
  readonly edge: string;
  readonly operator: TOperator;
}

export type TEdgeFilterValue =
  | IEdgeFilterValue<'eq', TFilterValue>
  | IEdgeFilterValue<'not', TFilterValue>;

interface IReverseEdgeFilterValue<TOperator extends string, TValue>
  extends IFilterValue<'ReverseEdge', TValue> {
  readonly reverseEdge: string;
  readonly operator: TOperator;
}

export type TReverseEdgeFilterValue =
  // Unique
  | IReverseEdgeFilterValue<'eq', TFilterValue>
  | IReverseEdgeFilterValue<'not', TFilterValue>

  // Non-unique
  | IReverseEdgeFilterValue<'none', TFilterValue>
  | IReverseEdgeFilterValue<'some', TFilterValue>
  | IReverseEdgeFilterValue<'every', TFilterValue>;

export type TFilterValue =
  | TBooleanFilterValue
  | TLogicalFilterValue
  | TLeafFilterValue
  | TEdgeFilterValue
  | TReverseEdgeFilterValue;
