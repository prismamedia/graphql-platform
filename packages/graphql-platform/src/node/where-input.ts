import {
  addPath,
  assertInputObject,
  GraphQLNonNullDecorator,
  isPublicEntry,
  Path,
  Public,
  UnexpectedValueError,
} from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import assert from 'assert';
import { GraphQLInputObjectType, GraphQLList, GraphQLNonNull } from 'graphql';
import { Node } from '../node';
import { Edge, Leaf, Scalars } from './component';
import { ReverseEdge } from './reverse-edge';
import { TFilterValue, TLeafFilterOperator } from './where-input/ast';
import { Filter } from './where-input/filter';
import { optimizeFilterValue } from './where-input/optimization';

export * from './where-input/ast';
export * from './where-input/filter';
export * from './where-input/optimization';

export type TWhereInputValue = { [filterName: string]: any };

export type TParsedWhereInputValue = TFilterValue;

export class WhereInput {
  public readonly public: boolean;
  public readonly name: string;
  public readonly description: string;

  public constructor(public readonly node: Node) {
    this.public = node.public;
    this.name = `${node.name}WhereInput`;
    this.description = `Used to filter the "${node.name}" nodes`;
  }

  protected getLeafFilters(leaf: Leaf): Filter[] {
    const filters: Filter[] = [];

    // eq
    if (leaf.isFilterableWith('eq')) {
      filters.push(
        new Filter(this, leaf.name, {
          public: leaf.public,
          type: leaf.type,
          nullable: leaf.nullable,
          parseValue: (value) => ({
            kind: 'Leaf',
            leaf: leaf.name,
            operator: 'eq',
            value,
          }),
        }),
      );
    }

    // not
    if (leaf.isFilterableWith('not')) {
      filters.push(
        new Filter(this, `${leaf.name}_not`, {
          public: leaf.public,
          type: leaf.type,
          nullable: leaf.nullable,
          parseValue: (value) => ({
            kind: 'Leaf',
            leaf: leaf.name,
            operator: 'not',
            value,
          }),
        }),
      );
    }

    // is_null
    if (
      leaf.isFilterableWith('eq') &&
      leaf.isFilterableWith('not') &&
      leaf.nullable
    ) {
      filters.push(
        new Filter(this, `${leaf.name}_is_null`, {
          public: leaf.public,
          type: Scalars.Boolean,
          parseValue: (value) => ({
            kind: 'Leaf',
            leaf: leaf.name,
            operator: value ? 'eq' : 'not',
            value: null,
          }),
        }),
      );
    }

    // in, not_in
    for (const operator of <TLeafFilterOperator[]>['in', 'not_in']) {
      if (leaf.isFilterableWith(operator)) {
        filters.push(
          new Filter(this, `${leaf.name}_${operator}`, {
            public: leaf.public,
            type: GraphQLList(
              GraphQLNonNullDecorator(leaf.type, !leaf.nullable),
            ),
            parseValue: (value: any) => ({
              kind: 'Leaf',
              leaf: leaf.name,
              operator,
              value,
            }),
          }),
        );
      }
    }

    // gt, gte, lt, lte
    for (const operator of <TLeafFilterOperator[]>['gt', 'gte', 'lt', 'lte']) {
      if (leaf.isFilterableWith(operator)) {
        filters.push(
          new Filter(this, `${leaf.name}_${operator}`, {
            public: leaf.public,
            type: leaf.type,
            parseValue: (value, path) => ({
              kind: 'Leaf',
              leaf: leaf.name,
              operator,
              value,
            }),
          }),
        );
      }
    }

    return filters;
  }

  protected getEdgeFilters(edge: Edge): Filter[] {
    const filters: Filter[] = [];

    // eq, not
    filters.push(
      new Filter(this, edge.name, {
        public: edge.public,
        type: () => edge.to.whereInput.type,
        nullable: edge.nullable,
        parseValue: (value, path) => ({
          kind: 'Edge',
          edge: edge.name,
          operator: 'eq',
          value: edge.to.whereInput.parseValue(value, path),
        }),
      }),

      new Filter(this, `${edge.name}_not`, {
        public: edge.public,
        type: () => edge.to.whereInput.type,
        nullable: edge.nullable,
        parseValue: (value, path) => ({
          kind: 'Edge',
          edge: edge.name,
          operator: 'not',
          value: edge.to.whereInput.parseValue(value, path),
        }),
      }),
    );

    // is_null
    if (edge.nullable) {
      filters.push(
        new Filter(this, `${edge.name}_is_null`, {
          public: edge.public,
          type: Scalars.Boolean,
          parseValue: (value, path) => ({
            kind: 'Edge',
            edge: edge.name,
            operator: 'eq',
            value: {
              kind: 'Boolean',
              value: !value,
            },
          }),
        }),
      );
    }

    return filters;
  }

  protected getReverseEdgeFilters(reverseEdge: ReverseEdge): Filter[] {
    const filters: Filter[] = [];

    if (reverseEdge.unique) {
      filters.push(
        new Filter(this, reverseEdge.name, {
          public: reverseEdge.public,
          type: () => reverseEdge.to.whereInput.type,
          parseValue: (value, path) => ({
            kind: 'ReverseEdge',
            reverseEdge: reverseEdge.name,
            operator: 'eq',
            value: reverseEdge.to.whereInput.parseValue(value, path),
          }),
        }),
        new Filter(this, `${reverseEdge.name}_not`, {
          public: reverseEdge.public,
          type: () => reverseEdge.to.whereInput.type,
          parseValue: (value, path) => ({
            kind: 'ReverseEdge',
            reverseEdge: reverseEdge.name,
            operator: 'not',
            value: reverseEdge.to.whereInput.parseValue(value, path),
          }),
        }),
        new Filter(this, `${reverseEdge.name}_is_null`, {
          public: reverseEdge.public,
          type: Scalars.Boolean,
          parseValue: (value, path) => ({
            kind: 'ReverseEdge',
            reverseEdge: reverseEdge.name,
            operator: 'eq',
            value: {
              kind: 'Boolean',
              value: !value,
            },
          }),
        }),
      );
    } else {
      filters.push(
        new Filter(this, `${reverseEdge.name}_none`, {
          public: reverseEdge.public,
          type: () => reverseEdge.to.whereInput.type,
          parseValue: (value, path) => ({
            kind: 'ReverseEdge',
            reverseEdge: reverseEdge.name,
            operator: 'none',
            value: reverseEdge.to.whereInput.parseValue(value, path),
          }),
        }),
        new Filter(this, `${reverseEdge.name}_some`, {
          public: reverseEdge.public,
          type: () => reverseEdge.to.whereInput.type,
          parseValue: (value, path) => ({
            kind: 'ReverseEdge',
            reverseEdge: reverseEdge.name,
            operator: 'some',
            value: reverseEdge.to.whereInput.parseValue(value, path),
          }),
        }),
        new Filter(this, `${reverseEdge.name}_every`, {
          public: reverseEdge.public,
          type: () => reverseEdge.to.whereInput.type,
          parseValue: (value, path) => ({
            kind: 'ReverseEdge',
            reverseEdge: reverseEdge.name,
            operator: 'every',
            value: reverseEdge.to.whereInput.parseValue(value, path),
          }),
        }),
      );
    }

    return filters;
  }

  protected getLogicalFilters(): Filter[] {
    return [
      new Filter(this, 'AND', {
        public: this.public,
        type: () => GraphQLList(GraphQLNonNull(this.type)),
        parseValue: (value, path) => {
          if (!Array.isArray(value)) {
            throw new UnexpectedValueError(value, `a [${this.name}!]`, path);
          }

          return {
            kind: 'Logical',
            operator: 'and',
            value: value.map((value, index) =>
              this.parseValue(value, addPath(path, index)),
            ),
          };
        },
      }),
      new Filter(this, 'OR', {
        public: this.public,
        type: () => GraphQLList(GraphQLNonNull(this.type)),
        parseValue: (value, path) => {
          if (!Array.isArray(value)) {
            throw new UnexpectedValueError(value, `a [${this.name}!]`, path);
          }

          return {
            kind: 'Logical',
            operator: 'or',
            value: value.map((value, index) =>
              this.parseValue(value, addPath(path, index)),
            ),
          };
        },
      }),
      new Filter(this, 'NOT', {
        public: this.public,
        type: () => this.type,
        parseValue: (value, path) => ({
          kind: 'Logical',
          operator: 'not',
          value: this.parseValue(value, path),
        }),
      }),
    ];
  }

  @Memoize()
  public get filterMap(): ReadonlyMap<string, Filter> {
    const filters: Filter[] = [];

    for (const component of this.node.componentMap.values()) {
      if (component instanceof Leaf) {
        filters.push(...this.getLeafFilters(component));
      } else {
        filters.push(...this.getEdgeFilters(component));
      }
    }

    for (const reverseEdge of this.node.reverseEdgeMap.values()) {
      filters.push(...this.getReverseEdgeFilters(reverseEdge));
    }

    filters.push(...this.getLogicalFilters());

    const filterMap = new Map<string, Filter>();

    for (const filter of filters) {
      if (filterMap.has(filter.name)) {
        throw new Error(
          `"${this.name}" contains at least 2 filters with the same name "${filter.name}"`,
        );
      }

      filterMap.set(filter.name, filter);
    }

    return filterMap;
  }

  @Memoize()
  public get publicFilterMap(): ReadonlyMap<string, Public<Filter>> {
    return new Map([...this.filterMap].filter(isPublicEntry));
  }

  @Memoize()
  public get type(): GraphQLInputObjectType {
    assert(this.public, `"${this.name}" is private`);

    assert(
      this.publicFilterMap.size > 0,
      `"${this.name}" expects at least one public filter`,
    );

    return new GraphQLInputObjectType({
      name: this.name,
      description: this.description,
      fields: () =>
        Object.fromEntries(
          Array.from(this.publicFilterMap.values(), (filter) => [
            filter.name,
            filter.graphqlInputFieldConfig,
          ]),
        ),
    });
  }

  public parseValue(
    value: TWhereInputValue | boolean | null | undefined,
    path: Path = addPath(undefined, this.name),
  ): TParsedWhereInputValue {
    if (value === null) {
      return {
        kind: 'Boolean',
        value: false,
      };
    } else if (typeof value === 'boolean') {
      return {
        kind: 'Boolean',
        value,
      };
    }

    const assertedValue = assertInputObject(
      value,
      this.filterMap.values(),
      path,
    );

    return optimizeFilterValue(
      {
        kind: 'Logical',
        operator: 'and',
        value: assertedValue
          ? Object.entries(assertedValue).map(([filterName, filterValue]) =>
              this.filterMap
                .get(filterName)!
                .parseValue(filterValue, addPath(path, filterName)),
            )
          : [],
      },
      this.node,
    );
  }
}
