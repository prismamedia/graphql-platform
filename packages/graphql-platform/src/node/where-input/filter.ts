import {
  InputField,
  InputFieldConfig,
  Path,
} from '@prismamedia/graphql-platform-utils';
import { Except } from 'type-fest';
import { WhereInput } from '../where-input';
import { TFilterValue } from './ast';

export interface FilterConfig
  extends Except<InputFieldConfig, 'required' | 'defaultValue'> {
  parseValue(value: any, path: Path): TFilterValue;
}

export class Filter extends InputField {
  public readonly parseValue: FilterConfig['parseValue'];

  public constructor(
    public readonly whereInput: WhereInput,
    name: string,
    { parseValue, ...config }: FilterConfig,
  ) {
    super(whereInput, name, { nullable: false, ...config, required: false });

    this.parseValue = parseValue.bind(this);
  }
}
