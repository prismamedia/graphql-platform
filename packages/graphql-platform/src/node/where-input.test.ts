import { GraphQLInputObjectType, printType } from 'graphql';
import { GraphQLPlatform } from '..';
import { MyGP, nodeNames, nodes } from '../__tests__/config';
import { TWhereInputValue, WhereInput } from './where-input';
import { TEdgeFilterValue, TFilterValue } from './where-input/ast';
import { isReferenceFilterValue } from './where-input/optimization';

describe('WhereInput', () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({ nodes });
  });

  it.each(nodeNames)('has a(n) %sWhereInput', (nodeName) => {
    const node = gp.getNode(nodeName);
    const whereInput = node.whereInput;

    expect(whereInput).toBeInstanceOf(WhereInput);
    expect(whereInput.public).toEqual(node.public);

    if (whereInput.public) {
      expect(whereInput.type).toBeInstanceOf(GraphQLInputObjectType);
      expect(
        printType(whereInput.type, { commentDescriptions: true }),
      ).toMatchSnapshot(whereInput.name);
    } else {
      expect(() => whereInput.type).toThrowError(
        `"${nodeName}WhereInput" is private`,
      );
    }
  });

  it.each([
    [
      'Article',
      { id_is_null: true },
      'At ArticleWhereInput - expects not to contain the extra field(s) "id_is_null", got: {"id_is_null":true}',
    ],
    [
      'Article',
      { id: null },
      `At ArticleWhereInput.id - expects a non-null value`,
    ],
    [
      'Article',
      { id: 123 },
      `At ArticleWhereInput.id - expects a valid "UUID", got: 123`,
    ],
    [
      'Article',
      { category: { parent: { title: null } } },
      `At ArticleWhereInput.category.parent.title - expects a non-null value`,
    ],
  ])(
    'throws an Error on %sWhereInput.parseValue(%p)',
    (nodeName, value, error) => {
      const node = gp.getNode(nodeName);
      const whereInput = node.whereInput;

      expect(() => whereInput.parseValue(value)).toThrowError(error);
    },
  );

  it.each<
    [string, TWhereInputValue | boolean | null | undefined, TFilterValue]
  >([
    ['Article', undefined, { kind: 'Boolean', value: true }],
    ['Article', null, { kind: 'Boolean', value: false }],
    ['Article', false, { kind: 'Boolean', value: false }],
    ['Article', true, { kind: 'Boolean', value: true }],

    // An empty "{}" is equal to "{ AND: [] }" is equal to "true"
    ['Article', {}, { kind: 'Boolean', value: true }],
    ['Article', { AND: [] }, { kind: 'Boolean', value: true }],

    // An empty "{ OR: [] }" is equal to "false"
    ['Article', { OR: [] }, { kind: 'Boolean', value: false }],

    // A simple leaf filter
    [
      'Article',
      { id: '94de8a4b-a25f-4659-ba13-c84761ef135b' },
      {
        kind: 'Leaf',
        leaf: 'id',
        operator: 'eq',
        value: '94de8a4b-a25f-4659-ba13-c84761ef135b',
      },
    ],

    // A simple edge filter
    [
      'Article',
      { category: { id: '66fd100c-1c47-469d-a89a-ce9a3fdb8470' } },
      {
        kind: 'Edge',
        edge: 'category',
        operator: 'eq',
        value: {
          kind: 'Leaf',
          leaf: 'id',
          operator: 'eq',
          value: '66fd100c-1c47-469d-a89a-ce9a3fdb8470',
        },
      },
    ],
    [
      'Article',
      // An article WITH any "category"
      { category: {} },
      {
        kind: 'Edge',
        edge: 'category',
        operator: 'eq',
        value: {
          kind: 'Boolean',
          value: true,
        },
      },
    ],
    [
      'Article',
      // An article WITHOUT any "category"
      { category: null },
      {
        kind: 'Edge',
        edge: 'category',
        operator: 'eq',
        value: {
          kind: 'Boolean',
          value: false,
        },
      },
    ],
    [
      'Article',
      // An article WITH any "createdBy" (as "createdBy" is non-null, we can remove this filter as it always true)
      { createdBy: {} },
      {
        kind: 'Boolean',
        value: true,
      },
    ],
    [
      'Article',
      // An article WITHOUT any "createdBy" (as "createdBy" is non-null, it is impossible) - as null is forbidden for a non-null edge, we use the optimisation of "{ OR: [] }" into "false"
      { createdBy: { OR: [] } },
      {
        kind: 'Boolean',
        value: false,
      },
    ],
    [
      'Article',
      { OR: [{ title: 'My title' }, { title: 'My title' }] },
      { kind: 'Leaf', leaf: 'title', operator: 'eq', value: 'My title' },
    ],
    ['Article', { NOT: {} }, { kind: 'Boolean', value: false }],
    ['Article', { NOT: { NOT: {} } }, { kind: 'Boolean', value: true }],
    [
      'Article',
      {
        createdAt_in: [
          new Date('2020-10-26T12:12:12.000Z'),
          new Date('2020-10-26T12:12:12.000Z'),
        ],
      },
      {
        kind: 'Leaf',
        leaf: 'createdAt',
        operator: 'eq',
        value: new Date('2020-10-26T12:12:12.000Z'),
      },
    ],
    [
      'Article',
      {
        _id: 123,
        title_not: 'My random title',
        body_is_null: false,
        OR: [
          { category_is_null: true },
          { category: { _id: 456 } },
          { category: { id: '367665fb-b29b-43d3-b6d0-ec70aaf2f703' } },
        ],
      },
      {
        kind: 'Logical',
        operator: 'and',
        value: [
          {
            kind: 'Leaf',
            leaf: '_id',
            operator: 'eq',
            value: 123,
          },
          {
            kind: 'Leaf',
            leaf: 'title',
            operator: 'not',
            value: 'My random title',
          },
          {
            kind: 'Leaf',
            leaf: 'body',
            operator: 'not',
            value: null,
          },
          {
            kind: 'Logical',
            operator: 'or',
            value: [
              {
                kind: 'Edge',
                edge: 'category',
                operator: 'eq',
                value: {
                  kind: 'Boolean',
                  value: false,
                },
              },
              {
                kind: 'Edge',
                edge: 'category',
                operator: 'eq',
                value: {
                  kind: 'Leaf',
                  leaf: '_id',
                  operator: 'eq',
                  value: 456,
                },
              },
              {
                kind: 'Edge',
                edge: 'category',
                operator: 'eq',
                value: {
                  kind: 'Leaf',
                  leaf: 'id',
                  operator: 'eq',
                  value: '367665fb-b29b-43d3-b6d0-ec70aaf2f703',
                },
              },
            ],
          },
        ],
      },
    ],
    [
      'Category',
      {
        OR: [
          {
            parent: null,
            order: 0,
          },
          {
            parent: null,
            slug: 'my-home',
          },
        ],
      },
      {
        kind: 'Logical',
        operator: 'or',
        value: [
          {
            kind: 'Logical',
            operator: 'and',
            value: [
              {
                kind: 'Edge',
                edge: 'parent',
                operator: 'eq',
                value: {
                  kind: 'Boolean',
                  value: false,
                },
              },
              {
                kind: 'Leaf',
                leaf: 'order',
                operator: 'eq',
                value: 0,
              },
            ],
          },
          {
            kind: 'Logical',
            operator: 'and',
            value: [
              {
                kind: 'Leaf',
                leaf: 'slug',
                operator: 'eq',
                value: 'my-home',
              },
              {
                kind: 'Edge',
                edge: 'parent',
                operator: 'eq',
                value: {
                  kind: 'Boolean',
                  value: false,
                },
              },
            ],
          },
        ],
      },
    ],
  ])('%sWhereInput.parseValue(%p) = %p', (nodeName, value, parsedValue) => {
    const node = gp.getNode(nodeName);
    const whereInput = node.whereInput;

    expect(whereInput.parseValue(value)).toEqual(parsedValue);
  });

  it.each<
    [nodeName: string, edgeFilterValue: TEdgeFilterValue, expectation: boolean]
  >([
    [
      'Article',
      {
        kind: 'Edge',
        edge: 'category',
        operator: 'eq',
        value: {
          kind: 'Leaf',
          leaf: 'id',
          operator: 'eq',
          value: '5287c3fb-73e9-4a1b-bc57-3cb5ffc2de61',
        },
      },
      false,
    ],
    [
      'Article',
      {
        kind: 'Edge',
        edge: 'category',
        operator: 'eq',
        value: {
          kind: 'Logical',
          operator: 'and',
          value: [
            {
              kind: 'Edge',
              edge: 'parent',
              operator: 'eq',
              value: {
                kind: 'Leaf',
                leaf: '_id',
                operator: 'eq',
                value: 123,
              },
            },
            {
              kind: 'Leaf',
              leaf: 'slug',
              operator: 'in',
              value: ['tv', 'sport', 'music'],
            },
          ],
        },
      },
      true,
    ],
  ])(
    'isReferenceFilterValue() detects when an edge filter is actually only a reference filter',
    (nodeName, edgeFilterValue, expectation) => {
      const node = gp.getNode(nodeName);

      expect(
        isReferenceFilterValue(
          edgeFilterValue,
          node.getEdge(edgeFilterValue.edge),
        ),
      ).toEqual(expectation);
    },
  );
});
