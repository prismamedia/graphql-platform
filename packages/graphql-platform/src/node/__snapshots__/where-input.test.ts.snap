// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhereInput has a(n) ArticleTagWhereInput: ArticleTagWhereInput 1`] = `
"# Used to filter the \\"ArticleTag\\" nodes
input ArticleTagWhereInput {
  article: ArticleWhereInput
  article_not: ArticleWhereInput
  tag: TagWhereInput
  tag_not: TagWhereInput
  order: NonNegativeInt
  order_not: NonNegativeInt
  order_in: [NonNegativeInt!]
  order_not_in: [NonNegativeInt!]
  order_gt: NonNegativeInt
  order_gte: NonNegativeInt
  order_lt: NonNegativeInt
  order_lte: NonNegativeInt
  AND: [ArticleTagWhereInput!]
  OR: [ArticleTagWhereInput!]
  NOT: ArticleTagWhereInput
}"
`;

exports[`WhereInput has a(n) ArticleWhereInput: ArticleWhereInput 1`] = `
"# Used to filter the \\"Article\\" nodes
input ArticleWhereInput {
  id: UUID
  id_not: UUID
  id_in: [UUID!]
  id_not_in: [UUID!]
  status: ArticleStatus
  status_not: ArticleStatus
  status_in: [ArticleStatus!]
  status_not_in: [ArticleStatus!]
  title: NonEmptyTrimmedString
  title_not: NonEmptyTrimmedString
  title_in: [NonEmptyTrimmedString!]
  title_not_in: [NonEmptyTrimmedString!]
  slug: NonEmptyTrimmedString
  slug_not: NonEmptyTrimmedString
  slug_in: [NonEmptyTrimmedString!]
  slug_not_in: [NonEmptyTrimmedString!]
  body: NonEmptyTrimmedString
  body_not: NonEmptyTrimmedString
  body_is_null: Boolean
  body_in: [NonEmptyTrimmedString]
  body_not_in: [NonEmptyTrimmedString]
  category: CategoryWhereInput
  category_not: CategoryWhereInput
  category_is_null: Boolean
  createdBy: UserWhereInput
  createdBy_not: UserWhereInput
  createdAt: DateTime
  createdAt_not: DateTime
  createdAt_in: [DateTime!]
  createdAt_not_in: [DateTime!]
  createdAt_gt: DateTime
  createdAt_gte: DateTime
  createdAt_lt: DateTime
  createdAt_lte: DateTime
  updatedBy: UserWhereInput
  updatedBy_not: UserWhereInput
  updatedAt: DateTime
  updatedAt_not: DateTime
  updatedAt_in: [DateTime!]
  updatedAt_not_in: [DateTime!]
  updatedAt_gt: DateTime
  updatedAt_gte: DateTime
  updatedAt_lt: DateTime
  updatedAt_lte: DateTime
  tags_none: ArticleTagWhereInput
  tags_some: ArticleTagWhereInput
  tags_every: ArticleTagWhereInput
  AND: [ArticleWhereInput!]
  OR: [ArticleWhereInput!]
  NOT: ArticleWhereInput
}"
`;

exports[`WhereInput has a(n) CategoryWhereInput: CategoryWhereInput 1`] = `
"# Used to filter the \\"Category\\" nodes
input CategoryWhereInput {
  id: UUID
  id_not: UUID
  id_in: [UUID!]
  id_not_in: [UUID!]
  title: NonEmptyTrimmedString
  title_not: NonEmptyTrimmedString
  title_in: [NonEmptyTrimmedString!]
  title_not_in: [NonEmptyTrimmedString!]
  slug: NonEmptyTrimmedString
  slug_not: NonEmptyTrimmedString
  slug_in: [NonEmptyTrimmedString!]
  slug_not_in: [NonEmptyTrimmedString!]
  parent: CategoryWhereInput
  parent_not: CategoryWhereInput
  parent_is_null: Boolean
  order: NonNegativeInt
  order_not: NonNegativeInt
  order_in: [NonNegativeInt!]
  order_not_in: [NonNegativeInt!]
  order_gt: NonNegativeInt
  order_gte: NonNegativeInt
  order_lt: NonNegativeInt
  order_lte: NonNegativeInt
  articles_none: ArticleWhereInput
  articles_some: ArticleWhereInput
  articles_every: ArticleWhereInput
  children_none: CategoryWhereInput
  children_some: CategoryWhereInput
  children_every: CategoryWhereInput
  AND: [CategoryWhereInput!]
  OR: [CategoryWhereInput!]
  NOT: CategoryWhereInput
}"
`;

exports[`WhereInput has a(n) HitWhereInput: HitWhereInput 1`] = `
"# Used to filter the \\"Hit\\" nodes
input HitWhereInput {
  _id: NonNegativeInt
  _id_not: NonNegativeInt
  _id_in: [NonNegativeInt!]
  _id_not_in: [NonNegativeInt!]
  _id_gt: NonNegativeInt
  _id_gte: NonNegativeInt
  _id_lt: NonNegativeInt
  _id_lte: NonNegativeInt
  at: DateTime
  at_not: DateTime
  at_in: [DateTime!]
  at_not_in: [DateTime!]
  at_gt: DateTime
  at_gte: DateTime
  at_lt: DateTime
  at_lte: DateTime
  AND: [HitWhereInput!]
  OR: [HitWhereInput!]
  NOT: HitWhereInput
}"
`;

exports[`WhereInput has a(n) TagWhereInput: TagWhereInput 1`] = `
"# Used to filter the \\"Tag\\" nodes
input TagWhereInput {
  id: UUID
  id_not: UUID
  id_in: [UUID!]
  id_not_in: [UUID!]
  title: NonEmptyTrimmedString
  title_not: NonEmptyTrimmedString
  title_in: [NonEmptyTrimmedString!]
  title_not_in: [NonEmptyTrimmedString!]
  slug: NonEmptyTrimmedString
  slug_not: NonEmptyTrimmedString
  slug_in: [NonEmptyTrimmedString!]
  slug_not_in: [NonEmptyTrimmedString!]
  createdAt: DateTime
  createdAt_not: DateTime
  createdAt_in: [DateTime!]
  createdAt_not_in: [DateTime!]
  createdAt_gt: DateTime
  createdAt_gte: DateTime
  createdAt_lt: DateTime
  createdAt_lte: DateTime
  updatedAt: DateTime
  updatedAt_not: DateTime
  updatedAt_in: [DateTime!]
  updatedAt_not_in: [DateTime!]
  updatedAt_gt: DateTime
  updatedAt_gte: DateTime
  updatedAt_lt: DateTime
  updatedAt_lte: DateTime
  articleTags_none: ArticleTagWhereInput
  articleTags_some: ArticleTagWhereInput
  articleTags_every: ArticleTagWhereInput
  AND: [TagWhereInput!]
  OR: [TagWhereInput!]
  NOT: TagWhereInput
}"
`;

exports[`WhereInput has a(n) UserProfileWhereInput: UserProfileWhereInput 1`] = `
"# Used to filter the \\"UserProfile\\" nodes
input UserProfileWhereInput {
  user: UserWhereInput
  user_not: UserWhereInput
  birthday: Date
  birthday_not: Date
  birthday_is_null: Boolean
  birthday_in: [Date]
  birthday_not_in: [Date]
  birthday_gt: Date
  birthday_gte: Date
  birthday_lt: Date
  birthday_lte: Date
  facebookId: ID
  facebookId_not: ID
  facebookId_is_null: Boolean
  facebookId_in: [ID]
  facebookId_not_in: [ID]
  googleId: ID
  googleId_not: ID
  googleId_is_null: Boolean
  googleId_in: [ID]
  googleId_not_in: [ID]
  twitterId: ID
  twitterId_not: ID
  twitterId_is_null: Boolean
  twitterId_in: [ID]
  twitterId_not_in: [ID]
  AND: [UserProfileWhereInput!]
  OR: [UserProfileWhereInput!]
  NOT: UserProfileWhereInput
}"
`;

exports[`WhereInput has a(n) UserWhereInput: UserWhereInput 1`] = `
"# Used to filter the \\"User\\" nodes
input UserWhereInput {
  id: UUID
  id_not: UUID
  id_in: [UUID!]
  id_not_in: [UUID!]
  username: NonEmptyTrimmedString
  username_not: NonEmptyTrimmedString
  username_in: [NonEmptyTrimmedString!]
  username_not_in: [NonEmptyTrimmedString!]
  created_none: ArticleWhereInput
  created_some: ArticleWhereInput
  created_every: ArticleWhereInput
  updated_none: ArticleWhereInput
  updated_some: ArticleWhereInput
  updated_every: ArticleWhereInput
  profile: UserProfileWhereInput
  profile_not: UserProfileWhereInput
  profile_is_null: Boolean
  AND: [UserWhereInput!]
  OR: [UserWhereInput!]
  NOT: UserWhereInput
}"
`;
